### PR TITLE
'type' attribute on HTMLInputElement should be stored as an Atom

### LINKED ISSUE #8180: 'type' attribute on HTMLInputElement should be stored as an Atom
https://reviewable.io/reviews/servo/servo/8178#-K1Q3RsxGBXZZgSAVFe2

### PR DESCRIPTION
**Pull request by frewsxcv**
Fixes #8180

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8249)

### CODE REVIEW

**Comment by nox** [comment]
This is correct, but the point of it is to then use the Atom directly where we currently use a bare string. Could you do it in that PR, or file a follow-up?

-S-awaiting-review +S-awaiting-answer 


---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8249)

**Comment by frewsxcv** [comment]
@nox Do you know any places in particular where it should get changed?

**Comment by nox** [comment]
@frewsxcv Your labelable stuff check whether an input element is `hidden`, doesn't it?

**Comment by nox** [comment]
```
git grep -F ".Type()"
```

**Comment by nox** [comment]
It is also probably used in layout.

**Comment by nox** [comment]
Ah no never mind layout.

**Comment by frewsxcv** [comment]
Should I create a method called `type_` which returns the atom?

**Comment by frewsxcv** [comment]
Updated

File: components/script/dom/htmlelement.rs [hunk 1]
```diff
@@ -345,7 +345,7 @@ impl HTMLElement {
             NodeTypeId::Element(ElementTypeId::HTMLElement(type_id)) =>
                 match type_id {
                     HTMLElementTypeId::HTMLInputElement =>
-                        self.downcast::<HTMLInputElement>().unwrap().Type() != "hidden",
+                        &*self.downcast::<HTMLInputElement>().unwrap().type_() != "hidden",
```
Comment (nox) [inline components/script/dom/htmlelement.rs]: `!= atom!("hidden")`?

Comment (frewsxcv) [inline components/script/dom/htmlelement.rs]: Is there a performance concern with my solution?

Comment (Ms2ger) [inline components/script/dom/htmlelement.rs]: The entire point of atoms is avoiding character-by-character string comparisons, so yes.

Comment (frewsxcv) [inline components/script/dom/htmlelement.rs]: Addressed in the latest force push

**Comment by nox** [comment]
bors-servo r+


---

Reviewed 2 of 2 files at r2, 5 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8249)

**Comment by nox** [comment]
@bors-servo r+

Fail.


---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8249)

**Comment by frewsxcv** [comment]
Lock file issues fixed

**Comment by frewsxcv** [comment]
Rebased.

**Comment by nox** [comment]
@bors-servo r+


---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8249)

**Pull request merged**